geo.go

```diff
@@ -0,0 +1,138 @@
+package tuna
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const IP2CUrl = "https://ip2c.org/"
+
+type Location struct {
+	IP          string `json:"IP"`
+	CountryCode string `json:"CountryCode"`
+	Country     string `json:"Country"`
+	City        string `json:"City"`
+}
+
+var emptyLocation = Location{}
+
+func (l *Location) Empty() bool {
+	if l == nil {
+		return true
+	}
+	return *l == emptyLocation
+}
+
+type IPFilter struct {
+	Allow    []Location `json:"Allow"`
+	Disallow []Location `json:"Disallow"`
+}
+
+func (f *IPFilter) Empty() bool {
+	if f == nil {
+		return true
+	}
+
+	for _, loc := range f.Allow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	for _, loc := range f.Disallow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (f *IPFilter) IPCheck(ip string) bool {
+	if f.Empty() {
+		return true
+	}
+	for _, loc := range f.Disallow {
+		if len(loc.IP) == 0 {
+			continue
+		}
+		if loc.IP == ip {
+			return false
+		}
+	}
+	for _, loc := range f.Allow {
+		if len(loc.IP) == 0 {
+			continue
+		}
+		if loc.IP == ip {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func (f *IPFilter) CountryCheck(code string) bool {
+	if f.Empty() || len(code) == 0 {
+		return true
+	}
+	for _, loc := range f.Disallow {
+		if len(loc.CountryCode) == 0 {
+			continue
+		}
+		if loc.CountryCode == code {
+			return false
+		}
+	}
+	for _, loc := range f.Allow {
+		if len(loc.CountryCode) == 0 {
+			continue
+		}
+		if loc.CountryCode == code {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func getLocationFromIP2C(ip string) (*Location, error) {
+	queryUrl := IP2CUrl + ip
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+	resp, err := client.Get(queryUrl)
+	if err != nil {
+		return &Location{}, err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
```

Comment:
Error here means ip2c service is down or network problem. It does not mean the input is invalid